### ISSUE TITLE
[Feature] Implement gradient checkpointing

### ISSUE THREAD

**Issue by zhouzaida** [open-mmlab/mmengine#1258]
### What is the feature?

`Gradient checkpointing` is a technique used to trade off memory usage for computation time during backpropagation. PyTorch provides `torch.utils.checkpoint.checkpoint` and `torch.utils.checkpoint.checkpoint_sequential` to implement the feature.
MMEngine aims to support this feature in a configurable manner.

### Any other context?

references

- https://pytorch.org/docs/stable/checkpoint.html
- https://github.com/prigoyal/pytorch_memonger/blob/master/tutorial/Checkpointing_for_PyTorch_models.ipynb
